src/base/locked_queue.h

```diff
@@ -83,28 +87,57 @@ class LockedQueue {
 
   // Returns disengaged object only when this queue is closed and empty, or pool
   // is shutting down.
-  Optional Pop(const WorkerPool& pool, ui32 shard = DEFAULT_SHARD) THREAD_SAFE {
+  Optional Pop(const WorkerPool& pool, const ui32 shard_queue_limit = NOT_STRICT_SHARDING,
+               const ui32 shard = DEFAULT_SHARD) THREAD_SAFE {
     DCHECK(timeout_ > Seconds::zero());
 
     UniqueLock lock(pop_mutex_);
+    if (shard_queue_limit == NOT_STRICT_SHARDING) {
+      return PopWithHint(pool, lock, shard);
+    } else {
+      return PopStrict(pool, lock,
+                       index_.MaybeOverloadedShard(shard_queue_limit, shard));
+    }
+  }
+
+ private:
+  Optional PopWithHint(const WorkerPool& pool,
+                       UniqueLock& lock,
+                       ui32 shard = DEFAULT_SHARD) THREAD_SAFE {
     do {
       pop_condition_.wait_for(lock, timeout_);
     } while (!closed_ && queue_.empty() && !pool.IsShuttingDown());
     if ((closed_ && queue_.empty()) || pool.IsShuttingDown()) {
       return Optional();
     }
+    return RemoveTaskFromQueue(index_.GetWithHint(shard, queue_.begin()));
+  }
 
-    auto it = index_.Get(shard, queue_.begin());
-    Optional&& obj = std::move(*it);
-    queue_.erase(it);
+  Optional PopStrict(const WorkerPool& pool,
+                     UniqueLock& lock,
+                     ui32 shard = DEFAULT_SHARD) THREAD_SAFE {
+    do {
+      index_.WaitForShard(shard, lock, timeout_);
```

Comment:
That's not good in case you've got overloaded shard. It may become not overloaded just before this call - and you will waste time for irrelevant shard.